pkcs11/pkcs11_test.go

```diff
@@ -594,3 +594,126 @@ func TestCreateCertificate(t *testing.T) {
 		t.Errorf("Returned certificate did not match loaded certificate")
 	}
 }
+
+func TestEncryptRSA(t *testing.T) {
+	msg := "Plain text to encrypt"
+	b := []byte(msg)
+	tests := []struct {
+		name string
+		bits int
+	}{
+		{"2048", 2048},
+		{"4096", 4096},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := newTestSlot(t)
+			o := keyOptions{RSABits: test.bits}
+			priv, err := s.generate(o)
+			if err != nil {
+				t.Fatalf("generate(%#v) failed: %v", o, err)
+			}
+			// SHA1 is the only hash function supported by softhsm
+			priv, err = WithHash(priv, crypto.SHA1)
+			if err != nil {
+				t.Errorf("WithHash error: %v", err)
+			}
+			rsaPriv, ok := priv.(*rsaPrivateKey)
+			if !ok {
+				t.Fatalf("Private Key unexpected type, got %T, want *rsaPrivateKey", priv)
+			}
+			_, err = rsaPriv.encryptRSA(b)
+			if err != nil {
+				t.Errorf("encryptRSA Error: %v", err)
+			}
+		})
+	}
+}
+
+func TestDecryptRSA(t *testing.T) {
+	msg := "Plain text to encrypt"
+	b := []byte(msg)
+	tests := []struct {
+		name string
+		bits int
+	}{
+		{"2048", 2048},
+		{"4096", 4096},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := newTestSlot(t)
+			o := keyOptions{RSABits: test.bits}
+			priv, err := s.generate(o)
+			if err != nil {
+				t.Fatalf("generate(%#v) failed: %v", o, err)
+			}
+			// SHA1 is the only hash function supported by softhsm
+			priv, err = WithHash(priv, crypto.SHA1)
+			if err != nil {
+				t.Errorf("WithHash error: %v", err)
+			}
+			rsaPriv, ok := priv.(*rsaPrivateKey)
+			if !ok {
+				t.Fatalf("Private Key unexpected type, got %T, want *rsaPrivateKey", priv)
+			}
+			cipher, err := rsaPriv.encryptRSA(b)
+			if err != nil {
+				t.Errorf("encryptRSA Error: %v", err)
+			}
+			decrypted, err := rsaPriv.decryptRSA(cipher)
+			if err != nil {
+				t.Errorf("decryptRSA Error: %v", err)
+			}
+			decrypted = bytes.Trim(decrypted, "\x00")
```

Comment:
This shouldn't be needed? Your callers definitely aren't going to do that. Any idea where the trailing \x00 comes from?

Reply:
I believe the padding is included in cDecryptedLen. I can update the decrypt function so that the caller does not have to trim it manually though. [PKCS11 Specification 5.2](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html#_Toc416959738:~:text=This%20number%20may%20somewhat%20exceed%20the%20precise%20number%20of%20bytes%20needed%2C%20but%20should%20not%20exceed%20it%20by%20a%20large%20amount.)

Reply:
That'd be great! Please add that link in a comment too :)